File: backend/app/models/payment_transaction.py
from datetime import UTC, datetime

from app.extensions import db
from app.models.base import UUIDMixin


class TransactionDirection:
    CREDIT = "CREDIT"
    DEBIT = "DEBIT"


class TransactionType:
    PAYMENT = "PAYMENT"
    REVERSAL = "REVERSAL"
    ADJUSTMENT = "ADJUSTMENT"


class PaymentTransaction(UUIDMixin, db.Model):
    __tablename__ = "payment_transactions"

    loan_id = db.Column(db.String(36), db.ForeignKey("loans.id"), nullable=False, index=True)
    amount = db.Column(db.Numeric(12, 2), nullable=False)
    direction = db.Column(db.String(10), nullable=False)
    transaction_type = db.Column(db.String(20), nullable=False)
    posted_at = db.Column(db.DateTime, nullable=False, default=lambda: datetime.now(UTC))
    payment_method = db.Column(db.String(50))
    notes = db.Column(db.Text)
    idempotency_key = db.Column(db.String(255), unique=True, nullable=True, index=True)
    created_by = db.Column(db.String(36), db.ForeignKey("users.id"))
    created_at = db.Column(db.DateTime, nullable=False, default=lambda: datetime.now(UTC))

    loan = db.relationship("Loan", foreign_keys=[loan_id])
    actor = db.relationship("User", foreign_keys=[created_by])
    allocations = db.relationship(
        "PaymentAllocation", back_populates="transaction", cascade="all, delete-orphan"
    )

File: frontend/Models/PaymentTransaction.cs
using System.Text.Json.Serialization;

namespace LendQ.Client.Models;

public class PaymentTransaction
{
    [JsonPropertyName("id")]
    public string Id { get; set; } = "";

    [JsonPropertyName("loan_id")]
    public string LoanId { get; set; } = "";

    [JsonPropertyName("amount")]
    public decimal Amount { get; set; }

    [JsonPropertyName("posted_at")]
    public DateTime PostedAt { get; set; }

    [JsonPropertyName("payment_method")]
    public string PaymentMethod { get; set; } = "";

    [JsonPropertyName("direction")]
    public string Direction { get; set; } = "";

    [JsonPropertyName("transaction_type")]
    public string TransactionType { get; set; } = "";

    [JsonPropertyName("notes")]
    public string? Notes { get; set; }

    [JsonPropertyName("allocations")]
    public List<PaymentAllocation>? Allocations { get; set; }
}
